examples/amm/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+use async_graphql::{Request, Response};
+use linera_sdk::{
+    base::{ContractAbi, ServiceAbi},
+    views::MapView,
+};
+use linera_views::views::ViewError;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub struct AmmAbi;
+
+impl ContractAbi for AmmAbi {
+    type InitializationArgument = ();
+    type Parameters = ();
+    type Operation = Operation;
+    type ApplicationCall = ();
+    type Message = ();
+    type SessionCall = ();
+    type Response = ();
+    type SessionState = ();
+}
+
+impl ServiceAbi for AmmAbi {
+    type Query = Request;
+    type QueryResponse = Response;
+    type Parameters = ();
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Operation {
+    Swap {
+        input_token: String,
+        output_token: String,
+        amount: u64,
+    },
+    AddLiquidity {
+        token: String,
+        amount: u64,
+    },
```

Comment:
Usually adding liquidity requires both tokens (at least for simple AMMs) in exactly the same ratio as the pool.

One way to make this more flexible is to swap the tokens that are being added to the pool in order to obtain the correct amount. The equation for how much to swap so that the resulting ratio matches the final pool ratio might be a bit tricky though.

In summary I'd suggest simply rejecting the operation if the ratio of the tokens to add doesn't match the current pool ratio (maybe within a margin of error, since rounding will need to be taken care of).